include/seqan3/search/algorithm/search_result_range.hpp

```diff
@@ -0,0 +1,331 @@
+#include <seqan3/std/concepts>
+#include <seqan3/std/ranges>
+
+namespace seqan3
+{
+
+/*!\brief An input range over the search results generated by the given search algorithm.
+ * \ingroup submodule_search_algorithm
+ * \implements std::ranges::input_range
+ *
+ * \tparam search_algorithm_t The type of the search algorithm; must model std::invocable with the value type of
+ *                            the given range query over queries.
+ * \tparam query_range_t The type of the range over the queries; must model std::ranges::forward_range.
+ *
+ * \details
+ *
+ * Provides a lazy input-range interface over the search results generated by the underlying search algorithm.
+ */
+template <typename search_algorithm_t, std::ranges::view query_range_t>
+//!\cond
+    requires std::ranges::forward_range<query_range_t> &&
+             std::invocable<search_algorithm_t, std::ranges::range_reference_t<query_range_t>>
+//!\endcond
+class search_result_range
+{
+private:
+    class search_result_range_iterator;
+
+    //!\brief Befriend the iterator type.
+    friend class search_result_range_iterator;
+
+    //!\brief The type of a single query.
+    using query_t = std::ranges::range_value_t<query_range_t>;
+    //!\brief The type of the buffer returned by invoking the search on a single query.
+    using hit_range_t = std::invoke_result_t<search_algorithm_t, query_t &>;
+    //!\brief The value type of the buffer returned by invoking the search on a single query.
+    using result_buffer_value_t = std::pair<size_t, std::ranges::range_value_t<hit_range_t>>;
+    //!\brief The type of the buffer used to store the hits.
+    using result_buffer_t = std::vector<result_buffer_value_t>;
+    //!\brief The wrapped query_range_t to add single pass behaviour.
+    using single_pass_query_range_t = decltype(views::zip(std::views::iota(0), std::declval<query_range_t &&>())
+                                             | views::single_pass_input);
+    //!\brief The iterator over the wrapped query range.
+    using single_pass_query_range_iterator = std::ranges::iterator_t<single_pass_query_range_t>;
+
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    search_result_range() = default; //!< Defaulted.
+    search_result_range(search_result_range const &) = delete; //!< This is a move-only type.
+    search_result_range(search_result_range &&) = default; //!< Defaulted.
+    search_result_range & operator=(search_result_range const &) = delete; //!< This is a move-only type.
+    search_result_range & operator=(search_result_range &&) = default; //!< Defaulted.
+    ~search_result_range() = default; //!< Defaulted.
+
+    /*!\brief Constructs a search result range.
+     *
+     * \param[in] search_algorithm The search algorithm to invoke on a single query.
+     * \param[in] query_range The range over the queries to be searched.
+     */
+    search_result_range(search_algorithm_t search_algorithm, query_range_t query_range) :
```

Comment:
What happens if `query_range_t` is a lvalue reference? (because of the move?)

Reply:
it never is since it is required to be a view.